### PR TITLE
[_native][5/8] Add general, column, and cross-CTA reductions

### PR DESCRIPTION
**Pull request by slayton58**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.18.0) (oldest at bottom):
* #195062
* #195061
* #195059
* __->__ #195051
* #195056
* #195052

Human Notes:

Complete axis coverage and add the multi-CTA path for large rows.

Codex:

Complete axis coverage on the shared reduction body. The general driver
represents TensorIterator geometry as runtime extent/stride pairs, allowing
one structural kernel to serve arbitrary dimensions and layouts. The column
driver vectorizes kept-axis outputs and splits tall reduced axes into
partials.

The cross-CTA driver serves few-row, large-N reductions with two fused
launches. Exact divisor splits use the optimized row body; the general
ragged split preserves coverage and global indices for prime lengths. The
dispatcher now routes row, column, cross-CTA, reduce-all, and arbitrary-axis
cases without adding a second general device kernel.

Review `tile.py` and `kernel_general.py` before `kernel_coltile.py` and
`kernel_xcta.py`.

Test Plan:

```
spin quicklint
python test/python_native/test_kernel_general.py
python test/python_native/test_kernel_xcta.py
python test/python_native/test_kernel_rowtile.py
python test/python_native/test_kernel_coltile.py
```

Authored with assistance from an AI coding agent.

### CODE REVIEW

**Comment by slayton58** [comment]
@claude /pr-review

**Comment by slayton58** [comment]
Review follow-up: 

There are no consumers here, and this kernel will be launched via. wrappers that will ensure the blocking failures cannot happen.

File: test/python_native/test_kernel_general.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import torch
+from torch.testing._internal.common_cuda import TEST_CUDA
+from torch.testing._internal.common_utils import run_tests, skipIfNoCuteDSL, TestCase
+
+
+@unittest.skipUnless(TEST_CUDA, "CUDA required")
+@skipIfNoCuteDSL
+class TestKernelGeneral(TestCase):
+    def test_reduce_dim_general_path(self):
+        import cutlass
```
Comment (drisspg) [inline test/python_native/test_kernel_general.py]: nit same inline import stuff

File: torch/_native/ops/reductions/kernel_general.py [hunk 2]
```diff
@@ -0,0 +1,530 @@
...
+def _probe(x, red_axes):
+    # A dummy output with the reduced dims set to 1, as reduce_op expects (it reads shapes and
+    # strides only). Shared by the classifier and the fallback so both see one TI decode.
+    return torch.empty(
+        [1 if i in red_axes else s for i, s in enumerate(x.shape)],
+        device=x.device,
+        dtype=x.dtype,
+    )
+
+
```
Comment (drisspg) [inline torch/_native/ops/reductions/kernel_general.py:196]: > AI-assisted review: Could we reuse a keepdim view of the output instead of allocating a CUDA tensor just to query TensorIterator's geometry? `_reduce` has already allocated `outs` before calling this, so the probe adds another output-sized allocation on every call, including plan-cache hits. For reductions with many kept elements this is substantial temporary device memory, despite only needing shape/stride metadata.

File: torch/_native/ops/reductions/kernel_general.py [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+    rem = cutlass.Int64(linear)
+    if npairs == 1:
+        return rem * vals[3]
+    off = cutlass.Int64(0)
+    for j in range(npairs - 1):
+        q = (rem * vals[4 * j]) >> vals[4 * j + 1]
+        off = off + (rem - q * vals[4 * j + 2]) * vals[4 * j + 3]
+        rem = q
+    return off + rem * vals[4 * (npairs - 1) + 3]
+
```
Comment (drisspg) [inline torch/_native/ops/reductions/kernel_general.py]: Semi nit; this type of indexing is very nice with layouts; I asume this is perhaps to mirrror aten but still

File: torch/_native/ops/reductions/kernel_general.py [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+        reduce_fn = trait.reduce  # local bind: attribute access trips a dyn loop
+        acc_dtype = trait.acc  # accumulator dtype (a compile-time Python class)
+        if const_expr(self.from_partials):
+            # Stage 2: COMBINE pre-reduced accumulator tuples from the per-field partial buffers. Each
+            # output's partials are a contiguous run, and the base must be decoded from kept_pairs or every
+            # row reads row 0's. The fold is a DYNAMIC full-wave loop plus a constexpr remainder, because a
+            # static unroll scaled compile time with the partial count (~3s at 1e5). Bind the trait's
+            # attributes to locals -- attribute access inside a dynamic loop trips the IR flattener.
+            combine_fn = trait.combine
+            fdtypes = trait.fdtypes
```
Comment (drisspg) [inline torch/_native/ops/reductions/kernel_general.py]: nit: I woudl trimthis commetns still further or just describe the flwo as a doc bloc and then comment the tage

File: torch/_native/ops/reductions/kernel_general.py [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+    out_str = it.element_strides(0)
+    red = [(it.shape[i], in_str[i]) for i in range(it.ndim) if out_str[i] == 0]
+    kept = [
+        (it.shape[i], in_str[i], out_str[i]) for i in range(it.ndim) if out_str[i] != 0
+    ]
+    kept.sort(key=lambda p: p[2])  # fastest output dim first
+    return red, [(e, s) for e, s, _ in kept]
+
+
+def _probe(x, red_axes):
```
Comment (drisspg) [inline torch/_native/ops/reductions/kernel_general.py]: can we type some fo these funcs? its ahrd to knw how they are used with the short vars

File: torch/_native/ops/reductions/kernel_general.py [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+
+
+def reduce_all(
+    trait, trait_key, x, out_dtype, block=_K0_ALL_BLOCK, grid_mult=_K0_ALL_GRID_MULT
+):
+    # Full-tensor reduce-all via the two-stage split: stage 1 grid-strides the flat input into G
+    # chunks, stage 2 folds them and projects once. Mirrors ATen's ctas_per_output and needs no
+    # reshape, so it serves any element count. Index traits are served too, since a single row
+    # makes the flat offset the column.
+    if not (x.is_cuda and x.is_contiguous()):
```
Comment (drisspg) [inline torch/_native/ops/reductions/kernel_general.py:425]: needs contigous?

File: torch/_native/ops/reductions/kernel_general.py [hunk 7]
```diff
@@ -0,0 +1,530 @@
...
+import torch
+from torch._tensor_iterator import reduce_op
+
+from ...cutedsl.dtypes import torch2cute
+from .._cutedsl import launch as _L
+from .._cutedsl.plan_cache import cached_plan
+from .._cutedsl.traits import block_reduce, WARP, warp_reduce
+
+
+def _magic(d):
```
Comment (drisspg) [inline torch/_native/ops/reductions/kernel_general.py]: https://docs.nvidia.com/cutlass/latest/media/docs/pythonDSL/cute_dsl_api/cute.html#cutlass.cute.FastDivmodDivisorV2
?

Comment (slayton58) [reply torch/_native/ops/reductions/kernel_general.py]: Pros:
* I don't maintain the code
* Uses 32b instead of 64b like current - this is guaranteed ok, and is faster

Con:
* Not available on 4.5.2 (which is what this stack was dev'd on), needs >= 4.6 -- I think that's fine, but does need a guard.

File: test/python_native/test_kernel_general.py (hunk 1 above)
Comment (slayton58) [reply test/python_native/test_kernel_general.py]: Fixed

File: torch/_native/ops/reductions/kernel_general.py (hunk 2 above)
Comment (slayton58) [reply torch/_native/ops/reductions/kernel_general.py:196]: This was attempted, ended up as a COW hazard (as implemented at least) -- claude explanation below:

> Tried this and had to back it out: it's a COW hazard. A keepdim view of x becomes TensorIterator's output operand, and TI takes an output's writable pointer, which materializes a copy-on-write
input — torch._lazy_clone(base) then torch.sum(x, dim=-1) left _is_cow_tensor(x) false. test_reductions_cutedsl.py::test_cow_input_served_and_preserved catches it. On the allocation itself:
at both call sites the probe runs before any output exists (the fast kernels allocate their own buffers downstream), so there's nothing already allocated to reuse. It stays an allocation, with
a comment saying why it must not alias x.

File: torch/_native/ops/reductions/kernel_general.py (hunk 3 above)
Comment (slayton58) [reply torch/_native/ops/reductions/kernel_general.py]: Partially to match aten, partial because my understanding is that layouts require static shape & stride, limiting how many cases a kernel can handle without recompile.

File: torch/_native/ops/reductions/kernel_general.py (hunk 4 above)
Comment (slayton58) [reply torch/_native/ops/reductions/kernel_general.py]: Done

File: torch/_native/ops/reductions/kernel_general.py (hunk 5 above)
Comment (slayton58) [reply torch/_native/ops/reductions/kernel_general.py]: These should be done now

File: torch/_native/ops/reductions/kernel_general.py (hunk 6 above)
Comment (slayton58) [reply torch/_native/ops/reductions/kernel_general.py:425]: Yes, and the consumer (in PR 11/11) this is enforced.

This is a specific set of reductions for now, post-this-stack it'll be extended to non-contiguous.

**Review by drisspg — APPROVED** [review]
